Readme.md
# Sinatra File CMS

This is a small Sinatra app built during RB175 course in Launch School.

It's a simple File CMS, that allows users to view, edit and add files.

To use the app, first add a new user in `users.yml` file, hashing the password using BCrypt.
This can be done inside of IRB.

The above is not necessary, but you won't be able to edit, add or delete files.

To run the app, navigate to the folder and simply run `ruby cms.rb`, the server will
listen on `localhost:4567`.

To run the tests, run `ruby test/cms_test.rb`.

data/changes.txt
These are some changes that happened.
Change in general is good and welcome.
But sometimes, a change can be stressful.

test/cms_test.rb
ENV["RACK_ENV"] = "test"

require "minitest/autorun"
require "rack/test"
require "minitest/reporters"

Minitest::Reporters.use!

require_relative '../cms'

class CMSTest < Minitest::Test
  include Rack::Test::Methods

  def app
    Sinatra::Application
  end

  def setup
    FileUtils.mkdir_p(data_path)
  end

  def teardown
    FileUtils.rm_rf(data_path)
  end

  def session
    last_request.env["rack.session"]
  end

  def admin_session
    { "rack.session" => { username: "admin" } }
  end

  def create_document(name, content = "")
    path = File.join(data_path, name)

    File.open(path, "w") do |file|
      file.write(content)
    end
  end

  def test_index
    create_document "about.md"
    create_document "changes.txt"
    create_document "history.txt"
    create_document "about.txt"

    get "/"

    assert_equal 200, last_response.status
    assert_equal "text/html;charset=utf-8", last_response["Content-Type"]
    assert_includes last_response.body, "about.txt"
    assert_includes last_response.body, "about.md"
    assert_includes last_response.body, "changes.txt"
    assert_includes last_response.body, "history.txt"
  end

  def test_history_text_document
    create_document "history.txt", "Ruby 0.95 released"

    get "/history.txt"

    assert_equal 200, last_response.status
    assert_equal "text/plain", last_response["Content-Type"]
    assert_includes last_response.body, "Ruby 0.95 released"
  end

  def test_nonexistent_document
    get "/wrong.txt"
    assert_equal 302, last_response.status
    assert_equal "wrong.txt does not exist.", session[:message]
  end

  def test_renders_markdown_document
    create_document "about.md", "# Ruby is..."

    get "/about.md"

    assert_equal 200, last_response.status
    assert_equal "text/html", last_response["Content-Type"]
    assert_includes last_response.body, "<h1>Ruby is...</h1>"
  end

  def test_document_can_be_edited
    time_now = Time.now.to_s # this represents dynamic text added to the file

    create_document "test.txt"

    get "/test.txt", {}, admin_session
    assert_equal 200, last_response.status
    refute_includes last_response.body, time_now # make sure the file doesn't contain the text

    get "/test.txt/edit"
    assert_equal 200, last_response.status
    assert_includes last_response.body, "Edit content of test.txt" # make sure edit page contains the name of the file
    assert_includes last_response.body, "<textarea" # make sure edit page contains the textarea

    post "/test.txt/edit", :file_content => time_now # make a post request, with the dynamic text to be added
    assert_equal 302, last_response.status

    get last_response["Location"] # follow the redirect
    assert_equal 200, last_response.status
    assert_includes last_response.body, "test.txt has been updated" # make sure "/" shows the flash message

    get "/test.txt" # go to the file content again
    assert_equal 200, last_response.status
    assert_includes last_response.body, time_now # make sure the file now DOES contain dynamic text
  end

  def test_new_document_page_renders
    get "/new", {}, admin_session

    assert_equal 200, last_response.status
    assert_includes last_response.body, "Add a new document:"
    assert_includes last_response.body, "<input"
    assert_includes last_response.body, "<button"
  end

  def test_new_document_can_be_created
    get "/test_document.txt", {}, admin_session
    assert_equal 302, last_response.status

    get last_response["Location"]
    assert_includes last_response.body, "test_document.txt does not exist."

    post "/new", :name => "test_document.txt"
    assert_equal 302, last_response.status

    get last_response["Location"]
    assert_equal 200, last_response.status
    assert_includes last_response.body, "test_document.txt was created."
  end

  def test_new_document_name_empty
    post "/new", {:name => "    "}, admin_session
    assert_equal 404, last_response.status
    assert_includes last_response.body, "A name is required."
  end

  def test_document_can_be_deleted
    create_document "to_delete.txt"

    get "/to_delete.txt/delete", {}, admin_session
    assert_equal 302, last_response.status

    get last_response["Location"]
    assert_equal 200, last_response.status
    assert_includes last_response.body, "to_delete.txt has been deleted."
  end

  def test_signin_page_renders
    get "/users/signin"

    assert_equal 200, last_response.status
    assert_includes last_response.body, "Username"
    assert_includes last_response.body, "Password"
    assert_includes last_response.body, "<input"
  end

  def test_signin_valid
    post "/users/signin", :username => "admin", :password => "secret"
    assert_equal 302, last_response.status

    get last_response["Location"]
    assert_includes last_response.body, "Welcome"
    assert_includes last_response.body, "Signed in as admin"
    assert_equal "admin", session[:username]
  end

  def test_signin_invalid_credentials
    post "/users/signin", :username => "guest", :password => "wrong"
    assert_equal 422, last_response.status
    assert_includes last_response.body, "Invalid Credentials"
  end

  def test_signout
    post "/users/signin", :username => "admin", :password => "secret"
    get last_response["Location"]
    assert_includes last_response.body, "Welcome!"

    post "/users/signout"
    get last_response["Location"]
    assert_includes last_response.body, "You have been signed out."
    assert_includes last_response.body, "Sign In"
    assert_nil session[:username]
  end

  def test_redirected_when_not_signed_in
    get "/new"
    assert_equal 302, last_response.status
    assert_equal "You must be signed in to do that.", session[:message]

    create_document "about.txt"
    get "/about.txt/edit"
    assert_equal 302, last_response.status
    assert_equal "You must be signed in to do that.", session[:message]
  end
end


cms.rb
require 'sinatra'
require 'sinatra/reloader'
require "redcarpet"
require 'yaml'
require 'bcrypt'

configure do
  enable :sessions
  set :session_secret, 'secret'
end

def data_path
  if ENV["RACK_ENV"] == "test"
    File.expand_path("../test/data", __FILE__)
  else
    File.expand_path("../data", __FILE__)
  end
end

def load_user_credentials
  credentials_path = if ENV["RACK_ENV"] == "test"
                       File.expand_path("../test/users.yml", __FILE__)
                     else
                       File.expand_path("../users.yml", __FILE__)
                     end
  YAML.load_file(credentials_path)
end

def render_markdown(file_path)
  markdown = Redcarpet::Markdown.new(Redcarpet::Render::HTML)
  markdown.render(File.read(file_path))
end

def render_text(file_path)
  File.read(file_path)
end

def load_file_content(file_path)
  case File.extname(file_path)
  when ".txt"
    headers["Content-Type"] = "text/plain"
    render_text(file_path)
  when ".md"
    headers["Content-Type"] = "text/html"
    erb render_markdown(file_path)
  else
    session[:message] = "This file type is not supported."
    redirect "/"
  end
end

def signed_in?(session)
  session[:username]
end

def check_signed_in(session, &block)
  if signed_in?(session)
    block.call
  else
    session[:message] = "You must be signed in to do that."
    redirect "/"
  end
end

def valid_credentials?(username, password)
  credentials = load_user_credentials

  if credentials.key?(username)
    BCrypt::Password.new(credentials[username]) == password
  else
    false
  end
end

# Shows all documents in the system
get "/" do
  pattern = File.join(data_path, "*")

  @files = Dir.glob(pattern).map do |path|
    File.basename(path)
  end

  erb :index
end

# Shows the new document form
get "/new" do
  check_signed_in(session) do
    erb :new_document
  end
end

# Creates a new document
post "/new" do
  check_signed_in(session) do
    file_name = params[:name].strip

    if file_name.size == 0
      session[:message] = "A name is required."
      status 404
      erb :new_document
    else
      file_path = File.join(data_path, file_name)

      File.write(file_path, "")
      session[:message] = "#{file_name} was created."

      redirect "/"
    end
  end
end

# Shows the content of the file
get "/:file" do
  file_path = File.join(data_path, params[:file])

  if File.exist? file_path
    load_file_content(file_path)
  else
    session[:message] = "#{params[:file]} does not exist."
    redirect "/"
  end
end

# Shows the edit document form
get "/:file/edit" do
  check_signed_in(session) do
    file_path = File.join(data_path, params[:file])

    if File.exist? file_path
      @content = File.read(file_path)

      erb :edit
    else
      session[:message] = "#{params[:file]} does not exist."
      redirect "/"
    end
  end
end

# Updates the documents contents
post "/:file/edit" do
  check_signed_in(session) do
    file_path = File.join(data_path, params[:file])
    file_content = params[:file_content]

    File.write(file_path, file_content)

    session[:message] = "#{params[:file]} has been updated."
    redirect "/"
  end
end

# Deletes a file
get "/:file/delete" do
  check_signed_in(session) do
    file_path = File.join(data_path, params[:file])

    File.delete(file_path)

    session[:message] = "#{params[:file]} has been deleted."
    redirect "/"
  end
end

# Displays a sign in form
get "/users/signin" do
  erb :signin
end

# Logs a user in if credentials are correct
post "/users/signin" do
  username = params[:username]
  password = params[:password]

  if valid_credentials?(username, password)
    session[:username] = username
    session[:message] = "Welcome!"
    redirect "/"
  else
    session[:message] = "Invalid Credentials"
    status 422
    erb :signin
  end
end

# Logs a user out and redirects to index
post "/users/signout" do
  session.delete(:username)
  session[:message] = "You have been signed out."

  redirect "/"
end
